extensions/amp-base-carousel/1.0/base-carousel.js

```diff
@@ -16,61 +16,82 @@
 import * as Preact from '../../../src/preact';
 import {Arrow} from './arrow';
 import {Scroller} from './scroller';
-import {toChildArray, useRef, useState} from '../../../src/preact';
+import {
+  toChildArray,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from '../../../src/preact';
 
 /**
- * @param {!JsonObject} props
+ * @param {!BaseCarouselProps} props
  * @return {PreactDef.Renderable}
  */
 export function BaseCarousel(props) {
   const {
-    'arrowPrev': arrowPrev,
-    'arrowNext': arrowNext,
-    'children': children,
-    'defaultSlide': defaultSlide,
-    'loop': loop,
-    ...rest
+    arrowPrev,
+    arrowNext,
+    children,
+    defaultSlide,
+    loop,
+    slide,
+    onSlideChange,
+    onLayout,
+    setAdvance,
   } = props;
   const childrenArray = toChildArray(children);
   const {length} = childrenArray;
-  const [curSlide, setCurSlide] = useState(defaultSlide || 0);
+  const [curSlide, setCurSlide] = useState(slide ? slide : defaultSlide || 0);
+  const slideState = slide ? slide : curSlide;
   const ignoreProgrammaticScroll = useRef(true);
-  const setRestingIndex = (i) => {
-    ignoreProgrammaticScroll.current = true;
-    setCurSlide(i);
-  };
-  const scrollRef = useRef(null);
   const advance = (dir) => {
     const container = scrollRef.current;
     // Modify scrollLeft is preferred to `setCurSlide` to enable smooth scroll.
     // Note: `setCurSlide` will still be called on debounce by scroll handler.
     container./* OK */ scrollLeft += container./* OK */ offsetWidth * dir;
   };
+  // Fire on first render when first slide is displayed.
+  useLayoutEffect(() => {
+    if (onLayout) {
+      onLayout();
+    }
+    if (setAdvance) {
+      setAdvance(advance);
+    }
+  }, [onLayout, setAdvance]);
+
+  const setRestingIndex = (i) => {
+    ignoreProgrammaticScroll.current = true;
+    setCurSlide(i);
+    if (onSlideChange) {
+      onSlideChange(i);
+    }
+  };
+  const scrollRef = useRef(null);
   const disableForDir = (dir) =>
-    !loop && (curSlide + dir < 0 || curSlide + dir >= length);
+    !loop && (slideState + dir < 0 || slideState + dir >= length);
   return (
-    <div {...rest}>
-      <Scroller
-        ignoreProgrammaticScroll={ignoreProgrammaticScroll}
-        loop={loop}
-        restingIndex={curSlide}
-        setRestingIndex={setRestingIndex}
-        scrollRef={scrollRef}
-      >
-        {childrenArray}
-      </Scroller>
-      <Arrow
-        customArrow={arrowPrev}
-        dir={-1}
-        disabled={disableForDir(-1)}
-        advance={() => advance(-1)}
-      />
-      <Arrow
-        customArrow={arrowNext}
-        dir={1}
-        disabled={disableForDir(1)}
-        advance={() => advance(1)}
-      />
+    <div>
+      {Scroller({
```

Comment:
JSX?

Reply:
I restored to JSX. To clarify the change originally was meant as an illustrative one for what it takes to silence the type errors as they stand now.